### PR TITLE
Fix function call args

### LINKED ISSUE #804: Wrapping is too greedy for function call with two object literal arguments
Input source code:

```javascript
deepCopyAndAsyncMapLeaves(
  { source: sourceValue, destination: destination[sourceKey] },
  { valueMapper, overwriteExistingKeys }
);
```

The output is still well-formatted [when I specify `printWidth: 80`][correct]:

```javascript
deepCopyAndAsyncMapLeaves(
  { source: sourceValue, destination: destination[sourceKey] },
  { valueMapper, overwriteExistingKeys }
);
```

But [with `printWidth: 100`][bad-looking], the output is harder to read:

```javascript
deepCopyAndAsyncMapLeaves({ source: sourceValue, destination: destination[sourceKey] }, {
  valueMapper,
  overwriteExistingKeys,
});
```

This 100-width formatting loses the 80-width version's property of clearly showing that there are two object arguments, and it doesn't even take up fewer lines of space. I would prefer the 80-width output even when width 100 is specified.

I am guessing this happens because arguments are greedily put onto the same line as the function call as long as they will fit. Instead, Prettier should look ahead at later arguments to see whether the argument list will be better balanced if it breaks that first argument onto a new line.

[correct]: https://prettier.github.io/prettier/#%7B%22content%22%3A%22deepCopyAndAsyncMapLeaves(%5Cn%20%20%7B%20source%3A%20sourceValue%2C%20destination%3A%20destination%5BsourceKey%5D%20%7D%2C%5Cn%20%20%7B%20valueMapper%2C%20overwriteExistingKeys%20%7D%5Cn)%3B%22%2C%22options%22%3A%7B%22printWidth%22%3A80%2C%22tabWidth%22%3A2%2C%22singleQuote%22%3Afalse%2C%22trailingComma%22%3Afalse%2C%22bracketSpacing%22%3Atrue%2C%22jsxBracketSameLine%22%3Afalse%2C%22doc%22%3Afalse%7D%7D

[bad-looking]: https://prettier.github.io/prettier/#%7B%22content%22%3A%22deepCopyAndAsyncMapLeaves(%5Cn%20%20%7B%20source%3A%20sourceValue%2C%20destination%3A%20destination%5BsourceKey%5D%20%7D%2C%5Cn%20%20%7B%20valueMapper%2C%20overwriteExistingKeys%20%7D%5Cn)%3B%22%2C%22options%22%3A%7B%22printWidth%22%3A100%2C%22tabWidth%22%3A2%2C%22singleQuote%22%3Afalse%2C%22trailingComma%22%3Afalse%2C%22bracketSpacing%22%3Atrue%2C%22jsxBracketSameLine%22%3Afalse%2C%22doc%22%3Afalse%7D%7D

### PR DESCRIPTION
**Pull request by yamafaktory**
This one should fix #804. It tried to implement what you suggested @vjeux and I hope this corresponds to what you were expecting!

### CODE REVIEW

File: src/printer.js [hunk 1]
```diff
@@ -1919,6 +1920,10 @@ function printArgumentsList(path, options, print) {
 
   if (groupLastArg) {
```
Comment (vjeux) [inline src/printer.js]: I was thinking of adding this check inside of `groupLastArg`.  We shouldn't group them if the last two have the same type.

**Comment by vjeux** [comment]
Can you add this as a test:

```js
function someFunction(url) { 
  return get(url)
    .then(
      json => dispatch(success(json)),
      error => dispatch(failed(error))
    );
}
```

https://github.com/prettier/prettier/issues/682#issuecomment-279605561

**Comment by vjeux** [comment]
```js
const mapChargeItems = fp.flow(
  l => l < 10 ? l: 1, 
  l => Immutable.Range(l).toMap()
);
```
https://github.com/prettier/prettier/issues/677#issuecomment-279217238

**Comment by vjeux** [comment]
```js
expect(new Range([0, 0], [0, 0])).toEqualAtomRange(new Range([0, 0], [0, 0]));
```

https://github.com/prettier/prettier/issues/253

**Comment by yamafaktory** [comment]
Tests added!

File: src/printer.js (hunk 1 above)
Comment (yamafaktory) [inline src/printer.js]: Do you think that I should move it? If so, is the naming `groupLastArg` still relevant?

Comment (vjeux) [inline src/printer.js]: at the end of the groupLastArg big condition add && (!penultimateArg || penultimateArg.type !== lastArg.type)

File: tests/break-calls/__snapshots__/jsfmt.spec.js.snap [hunk 2]
```diff
@@ -12,5 +44,36 @@ h(
...
+    error => dispatch(failed(error))
+  );
+}
+
+const mapChargeItems = fp.flow(
+  l => l < 10 ? l : 1,
+  l => Immutable.Range(l).toMap()
+);
+
+expect(new Range([0, 0], [0, 0])).toEqualAtomRange(new Range([0, 0], [0, 0]));
```
Comment (vjeux) [inline tests/break-calls/__snapshots__/jsfmt.spec.js.snap]: Can you add some characters such that it foesn't fit in a single lit

File: src/printer.js (hunk 1 above)
Comment (yamafaktory) [inline src/printer.js]: ok, so you want to keep the name, correct?

File: tests/break-calls/__snapshots__/jsfmt.spec.js.snap (hunk 2 above)
Comment (vjeux) [inline tests/break-calls/__snapshots__/jsfmt.spec.js.snap]: doesn't*
line*

**Comment by vjeux** [comment]
Awesome thanks! This solves so many issues :)

**Pull request merged**